In [34]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [79]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [80]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [81]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [82]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
market_scaled = StandardScaler().fit_transform(df_market_data[['price_change_percentage_24h', 'price_change_percentage_7d',
       'price_change_percentage_14d', 'price_change_percentage_30d',
       'price_change_percentage_60d', 'price_change_percentage_200d',
       'price_change_percentage_1y']])

# Diplay the first three rows of the scaled data
market_scaled[0:3]

array([[ 0.50852937,  0.49319307,  0.77220043,  0.23545963, -0.0674951 ,
        -0.35595348, -0.25163688],
       [ 0.18544589,  0.93444504,  0.55869212, -0.05434093, -0.27348273,
        -0.11575947, -0.19935211],
       [ 0.02177396, -0.70633685, -0.02168042, -0.06103015,  0.00800452,
        -0.55024692, -0.28206051]])

In [83]:
# Create a DataFrame with the scaled data

df_market_scaled = pd.DataFrame(market_scaled, columns=['price_change_percentage_24h', 'price_change_percentage_7d',
       'price_change_percentage_14d', 'price_change_percentage_30d',
       'price_change_percentage_60d', 'price_change_percentage_200d',
       'price_change_percentage_1y'])

df_market_scaled["coin_id"] = df_market_data.index

# Set the coinid column as index

df_market_scaled.set_index('coin_id',inplace=True)

# Display sample data

df_market_scaled.head()


price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Data.

In [84]:
# Create a list with the number of k-values from 1 to 11
k = list(range (2, 11))

In [85]:
# Create an empty list to store the inertia values

inertia = []

# Create a for loop to compute the inertia with each possible value of k
for i in k: # Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
    k_model = KMeans(n_clusters=i, random_state=0)
# 2. Fit the model to the data using `df_market_data_scaled`
    k_model.fit(df_market_scaled)
# 3. Append the model.inertia_ to the inertia list
    inertia.append(k_model.inertia_)


In [86]:
# Create a dictionary with the data to plot the Elbow curve

elbow_data = {"k": k, "inertia": inertia}

# Create the DataFrame from the elbow data
df_elbow = pd.DataFrame(elbow_data)

# Review the DataFrame
df_elbow.head()

# Create a DataFrame with the data to plot the Elbow curve

df_elbow_data = pd.DataFrame.from_dict(elbow_data)
df_elbow_data.head()

k     inertia
0  2  195.820218
1  3  123.190482
2  4   79.022435
3  5   65.302379
4  6   52.888518

In [87]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.

elbow_plot = df_elbow_data.hvplot.line(x = "k", y = "inertia",
    width=600,
    height=400,
    rot=90,
    title="Elbow Curve", 
    xticks=k
)
elbow_plot

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:**  The best K value is 4

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [88]:
# Initialize the K-Means model using the best value for k

k_model = KMeans(n_clusters=4, random_state=0)

In [89]:
# Fit the K-Means model using the scaled data

k_model.fit(df_market_scaled)

KMeans(n_clusters=4, random_state=0)

In [90]:
# Predict the clusters to group the cryptocurrencies using the scaled data

crypto_clusters = k_model.predict(df_market_scaled) 

# Print the resulting array of cluster values.

crypto_clusters

array([0, 0, 3, 3, 0, 0, 0, 0, 0, 3, 3, 3, 3, 0, 3, 0, 3, 3, 0, 3, 3, 0,
       3, 3, 3, 3, 3, 3, 0, 3, 3, 3, 1, 0, 3, 3, 2, 3, 3, 3, 3],
      dtype=int32)

In [91]:
# Create a copy of the DataFrame

market_data_predictions_df = df_market_scaled.copy()

In [92]:
# Add a new column to the DataFrame with the predicted clusters

market_data_predictions_df['crypto_segment'] = crypto_clusters

# Display sample data

market_data_predictions_df.head()


price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  crypto_segment  
coin_id                                                   
bitcoin                        -0.251637               0  
ethereum                       -0.199352               0  
tether                         -0.282061               3  
ripple                         -0.295546               3  
bitcoin-cash                   -0.270317               0

In [93]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.

cluster_plot = market_data_predictions_df.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by="crypto_segment",
    hover_cols = ["coin_id"], 
    title = "Scatter Plot by crypto clusters Segment - k=4"
)
cluster_plot


:NdOverlay   [crypto_segment]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [94]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)


In [95]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.

market_data_pca = pca.fit_transform(df_market_scaled)

# View the first five rows of the DataFrame. 

market_data_pca[:5]


array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527]])

In [96]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.

array_variance_ratio = pca.explained_variance_ratio_
array_variance_ratio


pca.explained_variance_ratio_


array([0.3719856 , 0.34700813, 0.17603793])

In [97]:
array_variance_ratio.sum()

0.8950316570309842

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** 0.8950316570309842 * 100 = 89.503166%

In [98]:
# Create a new DataFrame with the PCA data.
# Creating a DataFrame with the PCA data

df_market_data_pca = pd.DataFrame(market_data_pca, columns=["PC1", "PC2", "PC3"])

# Copy the crypto names from the original data

df_market_data_pca["coin_id"] = df_market_data.index

# Set the coinid column as index

df_market_data_pca = df_market_data_pca.set_index("coin_id")

# Display sample data

df_market_data_pca.head()

PC1       PC2       PC3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the PCA Data

In [99]:
# Create a list with the number of k-values from 1 to 11

k = list(range(1, 11))


In [100]:
# Create an empty list to store the inertia values

inertia= []

for i in k:   
# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters    
    model = KMeans(n_clusters=i, random_state= 0)
# 2. Fit the model to the data using `df_market_data_pca`
    model.fit(df_market_data_pca)
# 3. Append the model.inertia_ to the inertia list
    inertia.append(model.inertia_)

In [101]:
# Create a dictionary with the data to plot the Elbow curve

elbow_data_pca = {"k": k, "inertia": inertia}

# Create a DataFrame with the data to plot the Elbow curve

df_elbow_pca = pd.DataFrame(elbow_data_pca)

df_elbow_pca.head()

k     inertia
0  1  256.874086
1  2  165.901994
2  3   93.774626
3  4   49.665497
4  5   37.878747

In [102]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.

elbow_plot_pca = df_elbow_pca.hvplot.line(
    x="k",
    y="inertia",
    title="Elbow Curve",
    xticks=k
)
elbow_plot_pca

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** 4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** No, the value of k using the original data and PCA data is same.

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [103]:
# Initialize the K-Means model using the best value for k

model = KMeans(n_clusters=4, random_state=0)


In [104]:
# Fit the K-Means model using the PCA data
model.fit(df_market_data_pca)

KMeans(n_clusters=4, random_state=0)

In [105]:
# Predict the clusters to group the cryptocurrencies using the PCA data

k_model= model.predict(df_market_data_pca)

# Print the resulting array of cluster values.

k_model

array([0, 0, 3, 3, 0, 0, 0, 0, 0, 3, 3, 3, 3, 0, 3, 0, 3, 3, 0, 3, 3, 0,
       3, 3, 3, 3, 3, 3, 0, 3, 3, 3, 1, 0, 3, 3, 2, 3, 3, 3, 3],
      dtype=int32)

In [106]:
# Create a copy of the DataFrame with the PCA data

crypto_pca_predictions_df = df_market_data_pca.copy()

# Add a new column to the DataFrame with the predicted clusters

crypto_pca_predictions_df["pca_crypto_segments"] = k_model

# Display sample data

crypto_pca_predictions_df[:8]


PC1       PC2       PC3  pca_crypto_segments
coin_id                                                        
bitcoin      -0.600667  0.842760  0.461595                    0
ethereum     -0.458261  0.458466  0.952877                    0
tether       -0.433070 -0.168126 -0.641752                    3
ripple       -0.471835 -0.222660 -0.479053                    3
bitcoin-cash -1.157800  2.041209  1.859715                    0
binancecoin  -0.516534  1.388377  0.804071                    0
chainlink    -0.450711  0.517699  2.846143                    0
cardano      -0.345600  0.729439  1.478013                    0

In [107]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.


cluster_plot_pca = crypto_pca_predictions_df.hvplot.scatter(
    x="PC1",
    y="PC2",
    # Color the graph points with the labels found using K-Means and
    by="pca_crypto_segments",
    # add the crypto name in the `hover_cols` parameter to identify
    # the cryptocurrency represented by each data point.
    hover_cols=["coin_id"],
    title="Scatter Plot by Crypto Segment - k=4"

)
cluster_plot_pca

:NdOverlay   [pca_crypto_segments]
   :Scatter   [PC1]   (PC2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [108]:
# Composite plot to contrast the Elbow curves

elbow_plot + elbow_plot_pca

:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [109]:
# Composite plot to contrast the clusters
# YOUR CODE HERE!

cluster_plot + cluster_plot_pca


:Layout
   .NdOverlay.I  :NdOverlay   [crypto_segment]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [pca_crypto_segments]
      :Scatter   [PC1]   (PC2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** The elbow graphs and scatter plots remain unchanged, resulting in no impact as the optimal number of clusters for all cluster analyses remains at 4.
  